presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -46,4 +57,9 @@ default boolean isEnabled(Session session)
 
         Session getSession();
     }
+
+    interface Result
+    {
+        Result transformTo(PlanNode planNode);
```

Comment:
> I agree that thee two kind of methods you suggested (and even some more) will be eventually needed, but for now I would prefer not to pretend that something is possible while it is not.

This would enable us to write rules that will also target equivalence based optimizer, so we could do different efforts in parallel. Otherwise we would need to look at the rules (again) and adjust them to eq based optimizer.

> If you want I can rename this to replace.

That would be more explicit. Otherwise I guess we could keep only `replace` method, but we could already have more complete interface and target new rules against it.